### PR TITLE
Implement Error Handling

### LINKED ISSUE #184: Implement Error Handling
Currently we still use the default error handling of Spring. We should customize it:
* Hand out some error details (as they are not sensitive data)
* make sure that we get a proper 404 error when calling undefined routes
* make sure we get proper 400 errors when sending bad requests.

### PR DESCRIPTION
**Pull request by JonasKellerer**
Resolves #184

"make sure we get proper 400 errors when sending bad requests." is not included because we can not send bad requests yet and therefore can not be tested. Will be a new ticket.

### CODE REVIEW

File: lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    fun `throw NOT_FOUND(404) when route is not found`() {
+        every { lapisController.aggregated(any()) } returns AggregatedResponse(1)
+
+        mockMvc.perform(get("/notAValidRoute"))
+            .andExpect(status().isNotFound)
+    }
+
+    @Test
+    fun `throw INTERNAL_SERVER_ERROR(500) with additional info for any non specific error`() {
+        val validRoute: String = "/aggregated"
```
Comment (fengelniederhammer) [inline lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt:42]: We don't need the explicit type here

File: lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                      "message":"SomeMessage"
+                    }
+                    """.trimIndent(),
+                ),
+            )
+    }
+
+    @Test
+    fun `throw INTERNAL_SERVER_ERROR(500) with additional info for SiloExceptions`() {
+        val validRoute: String = "/aggregated"
```
Comment (fengelniederhammer) [inline lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt:62]: We don't need the explicit type here

File: lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt (hunk 1 above)
Comment (fengelniederhammer) [reply lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt:42]: Maybe make it a class constant? It's used in two test methods and will probably be used in further methods that we might add.

File: lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            .andExpect(status().isInternalServerError)
+            .andExpect(content().contentType(MediaType.APPLICATION_JSON))
+            .andExpect(
+                content().json(
+                    """
+                    {
+                      "title":"Unexpected error",
+                      "message":"SomeMessage"
+                    }
+                    """.trimIndent(),
```
Comment (fengelniederhammer) [inline lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt:55]: Seems like we don't need the trimIndent?
```suggestion
                    """,
```

File: lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            .andExpect(status().isInternalServerError)
+            .andExpect(content().contentType(MediaType.APPLICATION_JSON))
+            .andExpect(
+                content().json(
+                    """
+                    {
+                      "title":"Silo error",
+                      "message":"SomeMessage"
+                    }
+                    """.trimIndent(),
```
Comment (fengelniederhammer) [inline lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt:75]:
```suggestion
                    """,
```

File: lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    @BeforeEach
+    fun setUp() {
+        MockKAnnotations.init(this)
+    }
+
+    @Test
+    fun `throw NOT_FOUND(404) when route is not found`() {
+        every { lapisController.aggregated(any()) } returns AggregatedResponse(1)
+
+        mockMvc.perform(get("/notAValidRoute"))
```
Comment (fengelniederhammer) [inline lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt:36]: This doesn't look pretty, when calling it from a browser (The response differs depending on the `Accept` header). But I didn't find out yet why we can't see the behaviour here in the test...

**Review by fengelniederhammer — CHANGES_REQUESTED** [review]

File: lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt (hunk 5 above)
Comment (JonasKellerer) [reply lapis2/src/test/kotlin/org/genspectrum/lapis/controller/ExceptionHandlerTest.kt:41]: Now there should be a response, linking to our swagger-ui. However, testing was not possible yet. Either we can make the SpringBootTest working or we include it in the e2e tests

File: lapis2/src/main/resources/public/error/404.html [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Error 404</title>
+</head>
+<body>
+<h1>Page not found!</h1>
+<a href="/swagger-ui/index.html">Checkout our swagger-ui</a>
```
Comment (fengelniederhammer) [inline lapis2/src/main/resources/public/error/404.html:9]:
```suggestion
<a href="/swagger-ui/index.html">Visit our swagger-ui</a>
```

Sounds better IMO

**Pull request merged**